src/components/program/ProgramStageSelect.js

```diff
@@ -1,83 +1,85 @@
+import { useDataQuery } from '@dhis2/app-runtime'
 import i18n from '@dhis2/d2-i18n'
 import PropTypes from 'prop-types'
-import React, { Component } from 'react'
-import { connect } from 'react-redux'
-import { loadProgramStages } from '../../actions/programs.js'
+import React, { useCallback, useEffect } from 'react'
 import { SelectField } from '../core/index.js'
 
-class ProgramStageSelect extends Component {
-    static propTypes = {
-        loadProgramStages: PropTypes.func.isRequired,
-        onChange: PropTypes.func.isRequired,
-        className: PropTypes.string,
-        errorText: PropTypes.string,
-        program: PropTypes.object,
-        programStage: PropTypes.object,
-        programStages: PropTypes.object,
-    }
-
-    componentDidUpdate() {
-        const {
-            program,
-            programStage,
-            programStages,
-            loadProgramStages,
-            onChange,
-        } = this.props
+// Load all stages for one program
+const PROGRAM_STAGES_QUERY = {
+    stages: {
+        resource: 'programs',
+        id: ({ id }) => id,
+        params: {
+            fields: ['programStages[id,displayName~rename(name)]'],
+            paging: false,
+        },
+    },
+}
 
-        if (program) {
-            const stages = programStages[program.id]
-
-            // Load program stages when program is selected
-            if (!stages) {
-                loadProgramStages(program.id)
-            }
+const ProgramStageSelect = ({
+    program,
+    programStage,
+    onChange,
+    className,
+    errorText,
+}) => {
+    const onProgramStagesLoad = useCallback(
+        (data) => {
+            const stages = data.stages.programStages
 
             // Select first program stage if only one
-            if (program && !programStage && stages && stages.length === 1) {
+            if (!programStage && stages?.length === 1) {
                 onChange(stages[0])
             }
-        }
-    }
+        },
+        [programStage, onChange]
+    )
 
-    render() {
-        const {
-            program,
-            programStage,
-            programStages,
-            onChange,
-            className,
-            errorText,
-        } = this.props
-
-        if (!program) {
-            return null
+    // https://runtime.dhis2.nu/#/hooks/useDataQuery
+    const { loading, error, data, refetch } = useDataQuery(
+        PROGRAM_STAGES_QUERY,
+        {
+            variables: { id: program.id },
+            onComplete: onProgramStagesLoad,
+            lazy: true,
         }
+    )
 
-        let items = programStages[program.id]
+    // Fetch program stages when program is changed
+    useEffect(() => {
+        refetch({ id: program.id })
+    }, [program, refetch])
```

Comment:
technically you could use `program.id` in the dependency array, but it probably doesn't matter.

Reply:
Keeping program, but added program.id to propTypes